src/formatMarkup.js

```diff
@@ -39,28 +69,30 @@ export const diffableFormatter = function (markup) {
     let result = '\n' + '  '.repeat(indent) + '<' + node.nodeName;
 
     // Add attributes
-    if (node?.attrs?.length === 1) {
-      let attr = node?.attrs[0];
-      result = result + ' ' + attr.name + '="' + attr.value + '">';
-    } else if (node?.attrs?.length) {
+    const bracket = tagIsVoidElement && voidElementMode === 'xhtml' ? ' />' : '>';
+    if (!isElementNode || !node.attrs.length) {
+      result = result + bracket;
+    } else if (node.attrs.length === 1) {
+      let [attr] = node.attrs;
+      result = result + ' ' + attr.name + '="' + attr.value + '"' + bracket;
+    } else if (node.attrs.length > 1) {
       node.attrs.forEach((attr) => {
         result = result + '\n' + '  '.repeat(indent + 1) + attr.name + '="' + attr.value + '"';
       });
-      result = result + '\n' + '  '.repeat(indent) + '>';
-    } else {
-      result = result + '>';
+      result = result + '\n' + '  '.repeat(indent) + bracket.trim();
     }
 
     // Process child nodes
-    if (node.childNodes && node.childNodes.length) {
+    if (isElementNode) {
```

Comment:
The intent of the original `if` is a performance skip and safety check. Validate that the stuff exists, and there is a reason to do a loop.

The code was pretty clear as to what it's intent was.

I don't know what `isElementNode` means, or why it has to be checked before this loop. I can look to see that it equal to `defaultTreeAdapter.isElementNode(node)`, but I don't know what `defaultTreeAdapter.isElementNode` is either, because it is not documented. And I don't know what `defaultTreeAdapter` is, because.... it is not documented. The documentation page for it is just empty:

* https://parse5.js.org/variables/parse5.defaultTreeAdapter.html

I am now forced to go look at the source code of an HTML parser, and I'm not going to do that when this could just be handled with a simple JS `if` statement.

I have been doing this for enough years to know, YOU DO NOT TRUST 3RD PARTY APIs. When I update parse5 from v7.x.x to v9.x.x in 2 years, this code is all going to break.

The objects in the generated AST are complex, and volatile. The code written to interact with them should be as brain-dead simple as possible and use direct, straight-forward, conditional safety checks. I don't want to have to come back to this in 2 years, and have nothing work and then track down someone else's source code from 3-6 years ago, whenever that API was written, before it got deprecated or removed or whatever.

Reply:
Yes, the docs are a little wack, it _is_ documented, here:
- https://parse5.js.org/interfaces/parse5.TreeAdapter.html#isElementNode

Although, when you update from v7.x.x to v9.x.x the likelihood things will break is already high, but I get your point. Their code is exactly what you had before, but with a type predicate. If you are untrusting of depending on the third-party code, it can just be rewritten to keep things simple dependency-wise.

I've pulled the code in question into `src/formatMarkup.js`. This way you have ownership and visibility of these utilities.